src/pages/v2/p/[projectId].page.tsx

```diff
@@ -1,74 +1,138 @@
-import { NextRouter, useRouter } from 'next/router'
+import axios from 'axios'
+import { consolidateMetadata, ProjectMetadataV4 } from 'models/project-metadata'
+import { GetStaticPaths, GetStaticProps, InferGetStaticPropsType } from 'next'
 import { V2UserProvider } from 'providers/v2/UserProvider'
+import { paginateDepleteProjectsQueryCall } from 'utils/apollo'
+import { ipfsCidUrl } from 'utils/ipfs'
+import { V2ContractName } from 'models/v2/contracts'
+import { loadContract } from 'utils/contracts/loadContract'
+import { SEO } from 'components/common'
 
 import V2Dashboard from './components/V2Dashboard'
+import { readProvider } from 'constants/readProvider'
+import { readNetwork } from 'constants/networks'
+import { JUICEBOX_MONEY_METADATA_DOMAIN } from 'constants/v2/metadataDomain'
+import { V2_PROJECT_IDS } from 'constants/v2/projectIds'
 
-enum ProjectIdType {
-  ProjectId,
-  Handle,
+async function getMetadataCidFromContract(projectId: number) {
+  const network = readNetwork.name
+  const contract = await loadContract(
+    V2ContractName.JBProjects,
+    network,
+    readProvider,
+  )
+  if (!contract) {
+    throw new Error(`contract not found ${V2ContractName.JBProjects}`)
+  }
+  const metadataCid = (await contract.metadataContentOf(
+    projectId,
+    JUICEBOX_MONEY_METADATA_DOMAIN,
+  )) as string
+  return metadataCid
 }
 
-function getProjectIdOrHandle(
-  router: NextRouter,
-):
-  | { type: ProjectIdType.ProjectId; projectId: number }
-  | { type: ProjectIdType.Handle; handle: string }
-  | undefined {
-  const projectId = router.query.projectId as string | undefined
-  if (!projectId?.length) return
-
-  const isHandle = router.query.isHandle as string | undefined
-  if (isHandle && Boolean(isHandle)) {
-    return { type: ProjectIdType.Handle, handle: projectId }
+export const getStaticPaths: GetStaticPaths = async () => {
+  if (process.env.BUILD_CACHE_V2_PROJECTS === 'true') {
+    const projects = await paginateDepleteProjectsQueryCall({
+      variables: { where: { cv: '2' } },
+    })
+    const paths = projects.map(({ projectId }) => ({
+      params: { projectId: String(projectId) },
+    }))
+    return { paths, fallback: true }
   }
 
-  if (isNaN(parseInt(projectId))) {
-    return
+  return {
+    paths: [{ params: { projectId: String(V2_PROJECT_IDS.JUICEBOX_DAO) } }],
+    fallback: true,
   }
-
-  // Fallback to parsing as a project id
-  return { type: ProjectIdType.ProjectId, projectId: parseInt(projectId) }
 }
 
-export default function V2ProjectPage() {
-  const router = useRouter()
-  const projectIdOrHandle = getProjectIdOrHandle(router)
+export const getStaticProps: GetStaticProps<{
+  metadata: ProjectMetadataV4
+  projectId: number
+}> = async context => {
+  if (!context.params) throw new Error('params not supplied')
+  const projectId = parseInt(context.params.projectId as string)
+  const metadataCid = await getMetadataCidFromContract(projectId)
+  const url = ipfsCidUrl(metadataCid)
+  try {
```

Comment:
**suggestion** yeah maybe we can have a single fn for this fault-tolerant pinata fetch, and reuse here and V1.

Reply:
I was having a look at [this](https://www.npmjs.com/package/bottleneck) which I think is perfect

Reply:
Have implemented this 👍